### PR TITLE
fortify-headers: suppress stdio.h format warnings

### PR DESCRIPTION
**Pull request by ksperling-apple**
Prevent build failures when compiling packages with FORTIFY_SOURCE and -Werror=format-nonliteral by suppressing this warning within the fortified stdio.h. These are warnings (treated as errors) like
```
..._musl/include/fortify/stdio.h: In function 'snprintf':
..._musl/include/fortify/stdio.h:101:9: error: format not a string literal,
  argument types not checked [-Werror=format-nonliteral]
```
Some packages (e.g. glib2) currently work around this by disabling fortify-source entirely, which is not ideal (see e.g. https://github.com/openwrt/packages/issues/18884).

### CODE REVIEW

**Comment by ksperling-apple** [comment]
Other related issues: https://github.com/openwrt/packages/issues/18994 https://github.com/openwrt/packages/issues/19004

**Comment by ynezz** [comment]
> Some packages (e.g. glib2) currently work around this by disabling fortify-source entirely, which is not ideal (see e.g. https://github.com/openwrt/packages/issues/18884).

I've just tried to build `glib-2.74.0` with `CONFIG_PKG_FORTIFY_SOURCE_2=y` and removing `PKG_FORTIFY_SOURCE:=0` and it works fine, so it looks like this workaround is not needed at all?

**Comment by ynezz** [comment]
> Prevent build failures when compiling packages with FORTIFY_SOURCE and -Werror=format-nonliteral by suppressing this warning within the fortified stdio.h. 

IMO we should see those issues and ideally fix them, and using `PKG_FORTIFY_SOURCE:=0` as a temporary workaround.

**Comment by ksperling-apple** [comment]
@ynezz thanks for taking the time to look at this PR

> I've just tried to build glib-2.74.0 and libfido2-1.12.0 with CONFIG_PKG_FORTIFY_SOURCE_2=y and removing PKG_FORTIFY_SOURCE:=0 and it works fine, so it looks like this workaround is not needed at all?

I'm not sure if some global build flags defaults have changed since then around `-Wformat-nonliteral` or `-Werror`, but I'm definitely still hitting this in some custom packages I'm working on as well.

> IMO we should see those issues and ideally fix them, and using `PKG_FORTIFY_SOURCE:=0` as a temporary workaround.

These `format-nonliteral` warnings are not in the source of any of the specific packages being built, but within the fortify-headers wrappers of e.g. `snprintf`. So there are no issues to fix in the source packages themselves (if there was e.g. a non-literal format string in the source package, it wouldn't go away by turning off fortify either).

```
_FORTIFY_FN(snprintf) int snprintf(char *__s, size_t __n, const char *__f, ...)
{
	size_t __b = __builtin_object_size(__s, 0);

	if (__n > __b)
		__builtin_trap();
	return __orig_snprintf(__s, __n, __f, __builtin_va_arg_pack()); // <- format non-literal warning here
}
```

I have also tried annotating these wrappers with the correct `format` attributes, e.g.
```
_FORTIFY_FN(snprintf) __attribute__((format(printf, 3, 4))) int snprintf(char *__s, size_t __n, const char *__f, ...)
```
but this doesn't help either. Essentially this is some sort of issue with what GCC thinks it knows about the arguments of the wrapped (orig) vs wrapper `snprintf` and related functions. The only other way I've found to make these go away is to compile with `-ffreestanding` (which tells the compiler not to make any assumptions about the standard C functions like `snprintf`), but that would potentially have a much wider impact than suppressing a single warning (that's effectively of a false positive) in a small area of code.

So in summary I think
1. the issue is in fortify-headers, not any particular other package
2. suppressing the warnings just within these fortified wrapper functions is the most targeted / least risky fix.

**Comment by ynezz** [comment]
> I'm definitely still hitting this in some custom packages I'm working on as well.

Can you create [issue](https://github.com/openwrt/openwrt/issues) about that with all the details? Include as much info as possible (target, gcc version etc.), ideally add some smallest possible copy&paste-able reproducer.

> Essentially this is some sort of issue with what GCC thinks it knows about the arguments of the wrapped (orig) vs wrapper snprintf functions. 

Yes, it looks like something machine/target/GCC related and should be probably tracked down, maybe reported upstream and fixed properly.

> suppressing the warning just within these fortified wrapper functions is the most targeted / least risky fix

IMO those look to me like some rare/isolated issues, like in that case with [glib2 on powerpc64](https://github.com/openwrt/packages/pull/21467#issuecomment-1612694426), which should be handled at the right place. Adding such global workarounds in a security improving extension looks like we might risk missing/silencing some important warnings in the future.

**Comment by ksperling-apple** [comment]
> Can you create [issue](https://github.com/openwrt/openwrt/issues) about that with all the details? Include as much info as possible (target, gcc version etc.), ideally add some smallest possible copy&paste-able reproducer.

Will do, I'll try to narrow it down to a small example. This is on `mipsel_24kc` in my case.

> IMO those look to me like some rare/isolated issues, like in that case with [glib2 on powerpc64](https://github.com/openwrt/packages/pull/21467#issuecomment-1612694426), which should be handled at the right place. Adding such global workarounds in a security improving extension looks like we might risk missing/silencing some important warnings in the future.

Well, it does seem to have popped up in a number of packages so far, and missing out on fortify-source for entire packages seems worse security-wise than suppressing one warning for a handful of small inline functions, but I understand the discomfort with messing with this in a universally-used header.

**Comment by ksperling-apple** [comment]
@ynezz Bug with minimal repro is here: https://github.com/openwrt/openwrt/issues/13016
